File: app/src/main/java/com/example/myaac/util/GrammarEngine.kt
package com.example.myaac.util

import java.util.Locale

object GrammarEngine {

    private val pronouns = setOf("i", "you", "he", "she", "it", "we", "they", "me", "him", "her", "us", "them")
    private val verbs = setOf(
        "want", "like", "need", "love", "go", "see", "hear", "feel", "eat", "drink", 
        "sleep", "play", "work", "help", "stop", "come", "get", "take", "make", "do", "say"
    )
    private val adjectives = setOf(
        "happy", "sad", "mad", "hungry", "tired", "sick", "scared", "excited", 
        "good", "bad", "big", "small", "hot", "cold", "thirsty", "angry"
    )
    private val nouns = setOf(
        "apple", "banana", "cookie", "sandwich", "water", "milk", "juice", 
        "book", "crayon", "tablet", "bathroom", "home", "school", "mom", "dad",
        "toy", "ball", "car", "dog", "cat", "friend", "teacher"
    )

    private val articles = setOf("a", "an", "the")
    private val prepositions = setOf("to", "for", "with", "at", "by", "from", "in", "on")

    /**
     * Attempts to fix a list of words to be more grammatically correct.
     * Example: ["i", "want", "play"] -> "I want to play"
     */
    fun fixSentence(words: List<String>, languageCode: String = "en"): String {
        if (words.isEmpty()) return ""
        if (languageCode != "en" && languageCode != "he" && languageCode != "iw") {
            return words.joinToString(" ").replaceFirstChar { it.uppercase() }
        }

        // Basic Hebrew handling: join with spaces and capitalize (not really capitalization in Hebrew, but first char if English)
        if (languageCode == "he" || languageCode == "iw") {
             return words.joinToString(" ")
        }

        val processedWords = mutableListOf<String>()
        val input = words.flatMap { it.lowercase().split(" ") }.filter { it.isNotBlank() }

        if (input.isEmpty()) return ""

        for (i in input.indices) {
            val current = input[i]
            val next = input.getOrNull(i + 1)

            processedWords.add(current)

            if (next != null) {
                // Rule 1: Verb + Verb -> Insert "to"
                if (isVerb(current) && isVerb(next)) {
                    if (current != "stop" && current != "do" && current != "can" && next != "to") {
                        processedWords.add("to")
                    }
                }

                // Rule 2: Pronoun + Adjective -> Insert am/is/are
                if (isPronoun(current) && isAdjective(next)) {
                    val copula = when (current) {
                        "i" -> "am"
                        "you", "we", "they" -> "are"
                        "he", "she", "it" -> "is"
                        else -> null
                    }
                    if (copula != null && next != copula) {
                        processedWords.add(copula)
                    }
                }

                // Rule 3: Verb + Noun -> Insert article
                if (isVerb(current) && isNoun(next)) {
                   if (!isNonCountable(next) && !articles.contains(current) && !prepositions.contains(current)) {
                       val article = if ("aeiou".contains(next[0])) "an" else "a"
                       processedWords.add(article)
                   }
                }
            }
        }

        // Post-processing: remove duplicates if they happened (e.g. "want to to play")
        val finalWords = mutableListOf<String>()
        for (word in processedWords) {
            if (finalWords.isEmpty() || finalWords.last() != word) {
                finalWords.add(word)
            }
        }

        return finalWords.joinToString(" ") { word ->
            if (word == "i") "I" else word
        }.replaceFirstChar { it.uppercase() }
    }

    private fun isVerb(word: String) = verbs.contains(word)
    private fun isPronoun(word: String) = pronouns.contains(word)
    private fun isAdjective(word: String) = adjectives.contains(word)
    private fun isNoun(word: String) = nouns.contains(word)
    
    private fun isNonCountable(word: String) = setOf("water", "milk", "juice", "bread", "help", "work", "sleep").contains(word)
}

File: web/src/nlp/grammarEngine.ts
// Port of util/GrammarEngine.kt — near-mechanical translation.
// Rule-based sentence patcher (inserts "to"/articles/copulas between a bag of tapped words).

const PRONOUNS = new Set(['i', 'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'her', 'us', 'them']);
const VERBS = new Set([
  'want', 'like', 'need', 'love', 'go', 'see', 'hear', 'feel', 'eat', 'drink',
  'sleep', 'play', 'work', 'help', 'stop', 'come', 'get', 'take', 'make', 'do', 'say',
]);
const ADJECTIVES = new Set([
  'happy', 'sad', 'mad', 'hungry', 'tired', 'sick', 'scared', 'excited',
  'good', 'bad', 'big', 'small', 'hot', 'cold', 'thirsty', 'angry',
]);
const NOUNS = new Set([
  'apple', 'banana', 'cookie', 'sandwich', 'water', 'milk', 'juice',
  'book', 'crayon', 'tablet', 'bathroom', 'home', 'school', 'mom', 'dad',
  'toy', 'ball', 'car', 'dog', 'cat', 'friend', 'teacher', 'mall', 'park', 'store',
]);

const PLACES = new Set(['home', 'school', 'mall', 'park', 'store', 'bathroom']);

const ARTICLES = new Set(['a', 'an', 'the']);
const PREPOSITIONS = new Set(['to', 'for', 'with', 'at', 'by', 'from', 'in', 'on']);

const NON_COUNTABLE = new Set(['water', 'milk', 'juice', 'bread', 'help', 'work', 'sleep']);

export const GrammarEngine = {
  /**
   * Attempts to fix a list of words to be more grammatically correct.
   * Example: ["i", "want", "play"] -> "I want to play"
   */
  fixSentence(words: string[], languageCode: string = 'en'): string {
    if (words.length === 0) return '';
    if (languageCode !== 'en' && languageCode !== 'he' && languageCode !== 'iw') {
      const joined = words.join(' ');
      return capitalizeFirst(joined);
    }

    // Basic Hebrew handling: join with spaces and capitalize (not really capitalization in
    // Hebrew, but first char if English)
    if (languageCode === 'he' || languageCode === 'iw') {
      return words.join(' ');
    }

    const processedWords: string[] = [];
    const input = words
      .flatMap((w) => w.toLowerCase().split(' '))
      .filter((w) => w.trim().length > 0);

    if (input.length === 0) return '';

    for (let i = 0; i < input.length; i++) {
      const current = input[i];
      const next = input[i + 1];

      processedWords.push(current);

      if (next != null) {
        // Rule 1: Verb + Verb -> Insert "to"
        if (GrammarEngine.isVerb(current) && GrammarEngine.isVerb(next)) {
          if (current !== 'stop' && current !== 'do' && current !== 'can' && next !== 'to') {
            processedWords.push('to');
          }
        }

        // Rule 1.5: "go" + Place -> Insert "to"
        if ((current === 'go' || current === 'come') && GrammarEngine.isPlace(next)) {
          if (next !== 'home') {
            // "go home" is correct, "go to mall" is correct
            processedWords.push('to');
          }
        }

        // Rule 2: Pronoun + Adjective -> Insert am/is/are
        if (GrammarEngine.isPronoun(current) && GrammarEngine.isAdjective(next)) {
          const copula = ((): string | null => {
            switch (current) {
              case 'i':
                return 'am';
              case 'you':
              case 'we':
              case 'they':
                return 'are';
              case 'he':
              case 'she':
              case 'it':
                return 'is';
              default:
                return null;
            }
          })();
          if (copula != null && next !== copula) {
            processedWords.push(copula);
          }
        }

        // Rule 3: Verb + Noun -> Insert article
        if (GrammarEngine.isVerb(current) && GrammarEngine.isNoun(next)) {
          // Don't insert article if it's a place we just handled with "to", or if it's non-countable
          if (
            !GrammarEngine.isPlace(next) &&
            !isNonCountable(next) &&
            !ARTICLES.has(current) &&
            !PREPOSITIONS.has(current) &&
            next !== 'mom' &&
            next !== 'dad'
          ) {
            const article = 'aeiou'.includes(next[0]) ? 'an' : 'a';
            processedWords.push(article);
          }
        }
      }
    }

    // Post-processing: remove duplicates if they happened (e.g. "want to to play")
    const finalWords: string[] = [];
    for (const word of processedWords) {
      if (finalWords.length === 0 || finalWords[finalWords.length - 1] !== word) {
        finalWords.push(word);
      }
    }

    const joined = finalWords.map((word) => (word === 'i' ? 'I' : word)).join(' ');
    return capitalizeFirst(joined);
  },

  isVerb(word: string): boolean {
    return VERBS.has(word);
  },
  isPronoun(word: string): boolean {
    return PRONOUNS.has(word);
  },
  isAdjective(word: string): boolean {
    return ADJECTIVES.has(word);
  },
  isNoun(word: string): boolean {
    return NOUNS.has(word);
  },
  isPlace(word: string): boolean {
    return PLACES.has(word);
  },
};

function isNonCountable(word: string): boolean {
  return NON_COUNTABLE.has(word);
}

function capitalizeFirst(text: string): string {
  if (text.length === 0) return text;
  return text[0].toUpperCase() + text.slice(1);
}
